### PR TITLE
[Move] Sui Parsers for Module ID and SuiAddress

### PR DESCRIPTION
**Pull request by amnn**
## Description

Expose parsers for `SuiAddress` and `ModuleId` based on the `move-command-line-common` parser framework, for (later) use in GraphQL.

## Test Plan

Added new unit tests for these parsers:

```
sui-types$ cargo nextest run
```

### CODE REVIEW

File: external-crates/move/crates/move-command-line-common/src/types.rs [hunk 1]
```diff
@@ -24,9 +24,14 @@ pub enum TypeToken {
...
 #[derive(Eq, PartialEq, Debug, Clone)]
-pub struct ParsedStructType {
+pub struct ParsedModuleId {
     pub address: ParsedAddress,
-    pub module: String,
+    pub name: String,
+}
+
+#[derive(Eq, PartialEq, Debug, Clone)]
+pub struct ParsedStructType {
```
Comment (wlmyng) [inline external-crates/move/crates/move-command-line-common/src/types.rs:33]: Ok I see, so say we're looking at 0x2::coin::Coin, the new structure would be

ParsedStructType {
module: ParsedModuleId,
name: Coin,
type_args: Vec<T> // something like that
}

ParsedModuleId {
address: ParsedAddress,
name: coin
}

ParsedAddress remains unchanged

File: external-crates/move/crates/move-command-line-common/src/parser.rs [hunk 2]
```diff
@@ -136,18 +142,39 @@ impl<'a, Tok: Token, I: Iterator<Item = (Tok, &'a str)>> Parser<'a, Tok, I> {
...
     fn parse_type_impl(&mut self, depth: u64) -> Result<ParsedType> {
         self.count += 1;
 
         if depth > MAX_TYPE_DEPTH || self.count > MAX_TYPE_NODE_COUNT {
             bail!("Type exceeds maximum nesting depth or node count")
         }
-        let (tok, contents) = self.advance_any()?;
-        Ok(match (tok, contents) {
+
+        Ok(match self.advance_any()? {
```
Comment (wlmyng) [inline external-crates/move/crates/move-command-line-common/src/parser.rs:177]: this is functionally the same as what we have today right, just sugared up a bit?

File: external-crates/move/crates/move-command-line-common/src/parser.rs [hunk 3]
```diff
@@ -163,17 +190,11 @@ impl<'a, I: Iterator<Item = (TypeToken, &'a str)>> Parser<'a, TypeToken, I> {
...
-                };
-                let address = parse_address_impl(addr_tok, contents)?;
-                self.advance(TypeToken::ColonColon)?;
-                let module_contents = self.advance(TypeToken::Ident)?;
+
+            (tok @ (TypeToken::Ident | TypeToken::AddressIdent), contents) => {
+                let module = self.parse_module_id_impl(tok, contents)?;
                 self.advance(TypeToken::ColonColon)?;
-                let struct_contents = self.advance(TypeToken::Ident)?;
+                let name = self.advance(TypeToken::Ident)?.to_owned();
```
Comment (wlmyng) [inline external-crates/move/crates/move-command-line-common/src/parser.rs:197]: so we've parsed the address and module, and now looking at the Type?

File: external-crates/move/crates/move-command-line-common/src/parser.rs [hunk 4]
```diff
@@ -189,13 +210,12 @@ impl<'a, I: Iterator<Item = (TypeToken, &'a str)>> Parser<'a, TypeToken, I> {
                     _ => vec![],
                 };
                 ParsedType::Struct(ParsedStructType {
-                    address,
-                    module: module_contents.to_owned(),
-                    name: struct_contents.to_owned(),
+                    module,
+                    name,
```
Comment (wlmyng) [inline external-crates/move/crates/move-command-line-common/src/parser.rs:214]: clarification: having already parsed the module and type, we can construct ParsedStructType from them

**Review by wlmyng — APPROVED** [review]
ok lgtm!

File: external-crates/move/crates/move-command-line-common/src/parser.rs (hunk 2 above)
Comment (amnn) [reply external-crates/move/crates/move-command-line-common/src/parser.rs:177]: Yep!

File: external-crates/move/crates/move-command-line-common/src/parser.rs (hunk 3 above)
Comment (amnn) [reply external-crates/move/crates/move-command-line-common/src/parser.rs:197]: That's right

File: external-crates/move/crates/move-command-line-common/src/parser.rs (hunk 4 above)
Comment (amnn) [reply external-crates/move/crates/move-command-line-common/src/parser.rs:214]: Yes, this hasn't really changed, I just did this to make sure there's only one place where parsing of `ModuleId` is implemented, so now we treat a parsed struct type as a ModuleId, followed by `::[name]<[type-param],*>`.

**Review by tzakian — APPROVED** [review]
Changes look good to me!

File: crates/sui-types/src/lib.rs [hunk 5]
```diff
@@ -149,6 +150,18 @@ pub fn sui_framework_address_concat_string(suffix: &str) -> String {
     format!("{}{suffix}", SUI_FRAMEWORK_ADDRESS.to_hex_literal())
 }
```
Comment (stefan-mysten) [inline crates/sui-types/src/lib.rs:152]: nit: add some doc comments, particularly when one shoud or should not use this function.

**Comment by amnn** [comment]
Going to start staging the stack to land, in this PR!

**Pull request merged**